packages/daimo-api/src/contract/SwapClogMatcher.ts

```diff
@@ -0,0 +1,115 @@
+
+import { ForeignTokenTransfer } from "./foreignCoinIndexer";
+import { TokenRegistry } from "../server/tokenRegistry";
+import { retryBackoff } from "../utils/retryBackoff";
+
+export class SwapClogMatcher {
+  /** Map of transactionHash to transfers that are within that transaction */
+  private txHashToTransfers: Map<string, ForeignTokenTransfer[]> = new Map();
+
+  /** Map of starting "to" transfer to the corresponding "from" transfer */
+  private correspondingSwapTransfers: Map<
+    ForeignTokenTransfer,
+    ForeignTokenTransfer
+  > = new Map();
+
+  constructor(private tokenReg: TokenRegistry) {}
+
+  // Given a set transaction hashes, load any transfers that are associated
+  // with those transactions.
+  async loadSwapTransfers(
+    pg: Pool,
+    from: number,
+    to: number,
+    chainId: number,
+    transactionHashes: string[]
+  ) {
+    const startTime = Date.now();
+
+    const txHashes = transactionHashes.map((tx) => `\\x${tx.substring(2)}`);
+
+    const result = await retryBackoff(
+      `swapClogMatcher-logs-query-${from}-${to}`,
+      async () => {
+        return await pg.query(
+          `SELECT * from erc20_transfers 
+          WHERE (block_num BETWEEN $1 AND $2) 
+          AND chain_id = $3
+          AND tx_hash = ANY($4::bytea[]);`,
+          [from, to, chainId, txHashes as any]
+        );
+      }
+    );
+
+    for (const row of result.rows) {
+      const log = {
+        blockHash: bytesToHex(row.block_hash, { size: 32 }),
+        blockNumber: BigInt(row.block_num),
+        transactionHash: bytesToHex(row.tx_hash, { size: 32 }),
+        transactionIndex: row.tx_idx,
+        logIndex: row.log_idx,
+        address: getAddress(bytesToHex(row.log_addr, { size: 20 })),
+        from: getAddress(bytesToHex(row.f, { size: 20 })),
+        to: getAddress(bytesToHex(row.t, { size: 20 })),
+        value: BigInt(row.v),
+      };
+      const token = this.tokenReg.getToken(log.address, chainId, true); // include home coin
+      if (token == null) continue;
+
+      const foreignTransfer: ForeignTokenTransfer = {
+        ...log,
+        foreignToken: token,
+      };
+      this.txHashToTransfers.set(foreignTransfer.transactionHash, [
+        ...(this.txHashToTransfers.get(foreignTransfer.transactionHash) || []),
+        foreignTransfer,
+      ]);
+    }
+    console.log(
+      `[SWAP] loaded ${result.rows.length} bundled transfers in ${
+        Date.now() - startTime
+      }ms`
+    );
+  }
+
+  // For a SwapClog, find the corresponding "to" transfer given the "from"
+  // transfer (from a Daimo account to a foreign account).
+  getMatchingSwapTransfer(from: Address, transactionHash: string) {
+    const transfers = this.txHashToTransfers.get(transactionHash);
+    if (!transfers || transfers.length === 0) return null;
+
+    // Get the corresponding transfer to the given logIndex transfer.
+    let currentTransfer = transfers.find((t) => t.from === from);
```

Comment:
i recommend passing in the originating transfer (= the exact log coords, txHash+logIndex)

assuming there will only be one transfer from a given coin in the transaction  probably works today, but won't be robust in future. 4337 bundles many ops in one bundle tx.